libcrux-ml-kem/src/ntt.rs

```diff
@@ -1,71 +1,203 @@
 use crate::{
     hax_utils::hax_debug_assert,
-    polynomial::{PolynomialRingElement, VECTORS_IN_RING_ELEMENT, ZETAS_TIMES_MONTGOMERY_R},
+    polynomial::{get_zeta, PolynomialRingElement, VECTORS_IN_RING_ELEMENT},
     vector::{montgomery_multiply_fe, Operations},
 };
 
 #[inline(always)]
+#[hax_lib::fstar::options("--z3rlimit 200 --ext context_pruning")]
+#[hax_lib::fstar::before(
+    interface,
+    "[@@ \"opaque_to_smt\"]
+   let ntt_re_range_2 (#v_Vector: Type0)
+         {| i1: Libcrux_ml_kem.Vector.Traits.t_Operations v_Vector |}
+         (re: Libcrux_ml_kem.Polynomial.t_PolynomialRingElement v_Vector) =
+       forall (i:nat). i < 16 ==> Spec.Utils.is_i16b_array_opaque (11207+5*3328)
+            (Libcrux_ml_kem.Vector.Traits.f_to_i16_array (re.f_coefficients.[ sz i ]))"
+)]
+#[hax_lib::fstar::before(
+    interface,
+    "[@@ \"opaque_to_smt\"]
+    let ntt_re_range_1 (#v_Vector: Type0)
+            {| i1: Libcrux_ml_kem.Vector.Traits.t_Operations v_Vector |}
+            (re: Libcrux_ml_kem.Polynomial.t_PolynomialRingElement v_Vector) =
+        forall (i:nat). i < 16 ==> Spec.Utils.is_i16b_array_opaque (11207+6*3328)
+                (Libcrux_ml_kem.Vector.Traits.f_to_i16_array (re.f_coefficients.[ sz i ]))"
+)]
+#[hax_lib::requires(fstar!("v ${*zeta_i} == 63 /\\
+    ntt_re_range_2 $re"))]
+#[hax_lib::ensures(|result| fstar!("ntt_re_range_1 ${re}_future /\\
+    v ${*zeta_i}_future == 127"))]
 pub(crate) fn ntt_at_layer_1<Vector: Operations>(
     zeta_i: &mut usize,
     re: &mut PolynomialRingElement<Vector>,
     _layer: usize,
     _initial_coefficient_bound: usize,
```

Comment:
This doesn't actually seem to be used? Let's drop it.

Reply:
This is still open

Reply:
What doesn't seem to be used? I think the function and its annotations are all needed?

Reply:
After looking into it, let's leave these arguments in here.
The next round of proofs e.g. in this PR https://github.com/cryspen/libcrux/pull/713 and related branches will rely on these arguments, so it is convenient to leave them in here and the call-site arguments contain useful information.

Reply:
I removed one of these unused argument and added a comment for the other